akka-sample-sharding-scala/killrweather/src/main/scala/sample/killrweather/WeatherStation.scala

```diff
@@ -0,0 +1,132 @@
+    case object WindDirection extends DataType
+    case object OneHourPrecip extends DataType
+    case object SixHourPrecip extends DataType
+    case object SkyCondition extends DataType
+    val All: Set[DataType] = Set(Temperature, Dewpoint, Pressure, WindDirection, OneHourPrecip, SixHourPrecip, SkyCondition)
+  }
+
+  sealed trait Function
+  object Function {
+    case object HighLow extends Function
+    case object Average extends Function
+    case object Current extends Function
+    val All: Set[Function] = Set(HighLow, Average, Current)
+  }
+
+  /**
+   * Simplified weather measurement data, actual weather data event comprises many more data points.
+   *
+   * @param eventTime unix timestamp when collected
+   * @param dataType type of data
+   * @param value data point value
+   */
+  final case class Data(eventTime: Long, dataType: DataType, value: Double)
+  final case class TimeWindow(start: Long, end: Long, value: Double)
+
+
+  def apply(wsid: String): Behavior[Command] = running(wsid, Vector.empty)
+
+  private def average(values: Vector[Double]): Double =
+    if (values.isEmpty) Double.NaN
+    else values.sum / values.size
+
+  private def running(wsid: String, values: Vector[Data]): Behavior[WeatherStation.Command] =
+    Behaviors.setup { context =>
+      Behaviors.receiveMessage {
+        case Record(data, received, replyTo) =>
+          val updated = values :+ data
+          if (context.log.isDebugEnabled) {
+            val averageForSameType = average(updated.filter(_.dataType == data.dataType).map(_.value))
+            context.log.debugN("{} total readings from station {}, type {}, average {}, diff: processingTime - eventTime: {} ms",
+              updated.size,
+              data.dataType,
+              wsid,
+              averageForSameType,
+              received - data.eventTime
+            )
+          }
+          replyTo ! DataRecorded(wsid)
+          running(wsid, updated) // store
+
+        case Query(wsid, dataType, func, replyTo) =>
+          val valuesForType = values.filter(_.dataType == dataType)
+          val value =
+            func match {
+              case Function.Average =>
+                val start: Long = valuesForType.headOption.map(_.eventTime).getOrElse(0)
+                val end: Long = valuesForType.lastOption.map(_.eventTime).getOrElse(0)
+                Vector(TimeWindow(start, end, average(valuesForType.map(_.value))))
+
+              case Function.HighLow =>
+                val (start, min) = valuesForType.map(e => e.eventTime -> e.value).min
+                val (end, max) = valuesForType.map(e => e.eventTime -> e.value).max
+                Vector(TimeWindow(start, end, min), TimeWindow(start, end, max))
+
+              case Function.Current =>
+                Vector(valuesForType.lastOption
+                  .map(e => TimeWindow(e.eventTime, e.eventTime, e.value))
+                  .getOrElse(TimeWindow(0, 0, 0.0)))
+
+          }
```

Comment:
Changed it to actually use the datatype for filtering rather than listing them but then hardcoding usage of "temperature"